From f376dc40d0999ed25a59cf228db6cea00c6b2e0b Mon Sep 17 00:00:00 2001
From: Nathan Chancellor <nathan@kernel.org>
Date: Tue, 3 Feb 2026 01:41:10 -0700
Subject: [PATCH] lib/Kconfig.debug: Require a release version of LLVM 22 for
 context analysis

Android clang 22.0.1, the compiler used for android-mainline, is a
prerelease version of LLVM 22 that does not include the alias analysis
necessary to avoid many instances of -Wthread-safety-analysis. The
current CONFIG_WARN_CONTEXT_ANALYSIS dependencies allow any prerelease
version of LLVM 22 to enable the context analysis, regardless of whether
it actually has such support.

This was reasonable to do while LLVM 22 was in development so that
people could build from main and start testing and validating this in
their own code. Now that LLVM 22 is released as 22.1.0, upgrade the
check to require at least this version to ensure that a user's toolchain
actually has all the changes needed for a smooth experience with context
analysis.

Signed-off-by: Nathan Chancellor <nathan@kernel.org>
---
I plan to send this in three weeks or so when LLVM 21.1.0 is formally
released.
---
 lib/Kconfig.debug | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/lib/Kconfig.debug b/lib/Kconfig.debug
index 2b17f85513f9..dcaff85c366f 100644
--- a/lib/Kconfig.debug
+++ b/lib/Kconfig.debug
@@ -618,7 +618,7 @@ config DEBUG_FORCE_WEAK_PER_CPU
 
 config WARN_CONTEXT_ANALYSIS
 	bool "Compiler context-analysis warnings"
-	depends on CC_IS_CLANG && CLANG_VERSION >= 220000
+	depends on CC_IS_CLANG && CLANG_VERSION >= 220100
 	# Branch profiling re-defines "if", which messes with the compiler's
 	# ability to analyze __cond_acquires(..), resulting in false positives.
 	depends on !TRACE_BRANCH_PROFILING
@@ -629,7 +629,7 @@ config WARN_CONTEXT_ANALYSIS
 	  and releasing user-definable "context locks".
 
 	  Clang's name of the feature is "Thread Safety Analysis". Requires
-	  Clang 22 or later.
+	  Clang 22.1.0 or later.
 
 	  Produces warnings by default. Select CONFIG_WERROR if you wish to
 	  turn these warnings into errors.
-- 
2.52.0

